### ISSUE TITLE
[BUG] @Builder does not copy @JsonTypeIdResolver from Jackson

### ISSUE THREAD

**Issue by Nithanim** [projectlombok/lombok#3102]
**Describe the bug**
When using `@Builder` on a class with Jackson annotations, all annotations are copied correctly except `com.fasterxml.jackson.databind.annotation.JsonTypeIdResolver`.

**To Reproduce**
A runnable demonstration can be found here: https://github.com/Nithanim/lombok-jackson-builder-bug/blob/master/src/main/java/test/ItemInfo.java (Has main class and a JUnit test that calls it, so `mvn test` works.)

A slightly cut-down version you can find below, runnable with Jackson Databind and lombok on the classpath: 
```java
@Value
@Builder
public class ItemInfo {
  @JsonTypeInfo(
      use = JsonTypeInfo.Id.CUSTOM,
      include = JsonTypeInfo.As.EXTERNAL_PROPERTY,
      property = "type",
      visible = true)
  @JsonTypeIdResolver(ItemTypeResolver.class)
  String details;

  public static void main(String[] args) throws Exception {
    Field field = ItemInfo.class.getDeclaredField("details");
    Objects.requireNonNull(field.getAnnotation(JsonTypeInfo.class));
    Objects.requireNonNull(field.getAnnotation(JsonTypeIdResolver.class));

    Method method = ItemInfo.ItemInfoBuilder.class.getMethod("details", String.class);
    Objects.requireNonNull(method.getAnnotation(JsonTypeInfo.class), "JsonTypeInfo not on builder");
    // This fails
    Objects.requireNonNull(
        method.getAnnotation(JsonTypeIdResolver.class), "JsonTypeIdResolver not on builder");
  }

  public static class ItemTypeResolver extends TypeIdResolverBase {
    @Override
    public String idFromValue(Object o) {
      return null;
    }
    @Override
    public String idFromValueAndType(Object o, Class<?> aClass) {
      return null;
    }
    @Override
    public JsonTypeInfo.Id getMechanism() {
      return null;
    }
  }
}
```

**Expected behavior**
The `com.fasterxml.jackson.databind.annotation.JsonTypeIdResolver` is copied from the field to the builder method like other Jackson annotations.

Executing the test above no longer fails and the setter method in the generated builder class file is annotated with the same `JsonTypeIdResolver` as the field.

**Version info (please complete the following information):**
Lombok version: 1.18.22
Both older JDK 11
```
$ mvn --version
Apache Maven 3.6.3
Maven home: /usr/share/maven
Java version: 11.0.13, vendor: Ubuntu, runtime: /usr/lib/jvm/java-11-openjdk-amd64
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.15.16-xanmod1", arch: "amd64", family: "unix"
```
And Temurin JDK 17:
```
$ /home/nithanim/.jdks/temurin-17.0.1/bin/javac -version                                                                                
javac 17.0.1
$ /home/nithanim/.jdks/temurin-17.0.1/bin/java -version                                                                                 
openjdk version "17.0.1" 2021-10-19
OpenJDK Runtime Environment Temurin-17.0.1+12 (build 17.0.1+12)
OpenJDK 64-Bit Server VM Temurin-17.0.1+12 (build 17.0.1+12, mixed mode, sharing)
```

Executed from within IntelliJ:
* Tests directly
* Main class
* Maven `test` target

**Additional context**
Skipping through lombok source I found this: https://github.com/projectlombok/lombok/blob/master/src/core/lombok/core/handlers/HandlerUtil.java#L445 It might be enough to add it there?
(Reached via https://github.com/projectlombok/lombok/blob/5120abe4741c78d19d7e65404f407cfe57074a47/src/core/lombok/javac/handlers/JavacHandlerUtil.java#L1570 and https://github.com/projectlombok/lombok/blob/5120abe4741c78d19d7e65404f407cfe57074a47/src/core/lombok/javac/handlers/HandleBuilder.java#L257)

Thank you!

**Comment by rzwitserloot** [comment]
Yeah, it's enough to add it in there. The only question is: Is it appropriate to copy annotation `com.fasterxml.jackson.databind.annotation.JsonTypeIdResolver` __in all circumstances regardless of context__?

I think the answer is 'yes', and then it is indeed as simple as adding it.

**Comment by Nithanim** [comment]
Thank you for your support!

In the given circumstance when deserializing we land here: https://github.com/FasterXML/jackson-databind/blob/2.14/src/main/java/com/fasterxml/jackson/databind/introspect/JacksonAnnotationIntrospector.java#L1493 Given that both annotations complement each other and that JsonTypeInfo is already in the whitelist in Lombok, `com.fasterxml.jackson.databind.annotation.JsonTypeIdResolver` should be treated equally. 

However the above solution has a mistake. It probably should be rather added to `COPY_TO_SETTER_ANNOTATIONS` instead of `BASE_COPYABLE_ANNOTATIONS` because the right place would be there: https://github.com/projectlombok/lombok/blob/master/src/core/lombok/javac/handlers/HandleBuilder.java#L924 (also the other annotations are there).
Otherwise, with `BASE_COPYABLE_ANNOTATIONS` the annotation is attached to the parameter instead of the setter method of the builder.

I am not sure if the annotation should then also be added to `JACKSON_COPY_TO_BUILDER_ANNOTATIONS`.

**Comment by janrieke** [comment]
Adding it to `JACKSON_COPY_TO_BUILDER_ANNOTATIONS` should not be necessary. I strongly suspect Jackson will perform type identification before it will determine in which way it should instantiate. So this annotation will probably have no effect on the builder class. However, I did not test it (and I don't know when I have time to do so).

**Comment by Nithanim** [comment]
Jackson cannot know the type beforehand so it has to discover the type while deserializing. That is the whole point of [TypeIdResolver](https://fasterxml.github.io/jackson-databind/javadoc/2.7/com/fasterxml/jackson/databind/jsontype/TypeIdResolver.html). JsonTypeInfo is the descriptor where to find the id (extracted from the parent's field in my case) and passes it to the TypeIdResolver to find the appropriate class. Currently, only the JsonTypeInfo is found, so Jackson has no idea what to do.

I have updated the JUnit test in https://github.com/Nithanim/lombok-jackson-builder-bug/tree/d07d2e82e812576ab1cc336bc89b3ce6fdf9e48b

**Comment by janrieke** [comment]
`JACKSON_COPY_TO_BUILDER_ANNOTATIONS` means "copy these annotations from the class to the builder class. It does not affect copying annotations from fields to builder setter methods.

**Comment by Nithanim** [comment]
Oh, sorry! I now know what you mean (I think). My reasoning was, that the builder will be used for discovery instead of the annotated class. But you are probably right that that does not make any sense.
